### ISSUE TITLE
build: Exercise.output is not consistent

### ISSUE THREAD

**Issue by perillo** [ratfactor/ziglings#248]
I'm adding a test for `zig build` in order to check that all exercises are processes in the correct order.
For each exercise, I check that:
  1. The first line is `Compiling xxx_exercise.zig...`
  2. The second line is `Checking xxx_exercise.zig...`
  3. The third line is `PASSED:`

Finally, I skip all the lines from the exercise output.  The number of lines to skip is the count of `\n` in `Exercise.output` + `1`.
The problem is that exercises `068_comptime3.zig` and `099_formatting.zig` broke the test.

There are two solutions:
  1. Add the actual content of the exercise output to `Exercise.output`
  2. Remove the extra LF in the two tests.

What do you think is the best solution?

P.S.:
Exercise `099_formatting.zig` is a special case, because after each row it uses double line feed.  This means that it will still contain a trailing LF in case the solution 2 is chosen.

**Comment by chrboesch** [comment]
I don't quite follow you, when I run 'eowyn' I get the following output:
```
...
Compiling 067_comptime2.zig...
Checking 067_comptime2.zig...
PASSED:
A BB CCC DDDD

Compiling 068_comptime3.zig...
Checking 068_comptime3.zig...
PASSED:
Minnow (1:32, 4 x 2)
Shark (1:16, 8 x 5)
Whale (1:1, 143 x 95)

Compiling 069_comptime4.zig...
Checking 069_comptime4.zig...
PASSED:
s1={ 1, 2, 3 }, s2={ 1, 2, 3, 4, 5 }, s3={ 1, 2, 3, 4, 5, 6, 7 }

...

Compiling 099_formatting.zig...
Checking 099_formatting.zig...
PASSED:

 X |  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15 
---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
 1 |  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15 

 2 |  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30 

 3 |  3   6   9  12  15  18  21  24  27  30  33  36  39  42  45 

 4 |  4   8  12  16  20  24  28  32  36  40  44  48  52  56  60 

 5 |  5  10  15  20  25  30  35  40  45  50  55  60  65  70  75 
...
```
What do you think is wrong with that?

**Comment by perillo** [comment]
The actual output is correct.  The problem is `Exercise.output`.

All exercises have a line feed at the end of the output, but `Exercise.output` does not have it, except for `068_comptime3.zig` and `099_formatting.zig`.
 
The check done in `ZiglingStep` does not fail, because the implementation do:
```zig
const trimOutput = std.mem.trimRight(u8, output, " \r\n");
const trimExerciseOutput = std.mem.trimRight(u8, self.exercise.output, " \r\n");
```

In the test I need to know how many lines are in the exercise output.

**Comment by chrboesch** [comment]
Ok, I hope now I have understand you. I tried it out and it works. Solves it this?

**Comment by perillo** [comment]
> Ok, I hope now I have understand you. I tried it out and it works. Solves it this?

I have commented on the PR.

Note that another solution is to make `Exercise.output` contain the *exact* bytes generated by the exercise **and** update `ZiglingStep` to do a strict comparison, without stripping whitespace at the end.

**Comment by chrboesch** [comment]
As far as I can remember, one of the reasons for this "soft" compare was that there were problems on certain systems, see: https://github.com/ratfactor/ziglings/pull/145
So if you want to change the `exercise.output`, what improvement will that bring?

**Comment by perillo** [comment]
Here is a draft of `test/tests.zig`, that finally add a test for `zig build -Dhealed` and `zig build -Dhealed -Dn=1 start`.  The test tests that each exercise are processed in the correct order:

https://gist.github.com/perillo/6ca5f3f9ea5c502d71a57f95bdd098ba

The function doing the check is:
https://gist.github.com/perillo/6ca5f3f9ea5c502d71a57f95bdd098ba#file-tests-zig-L196

The `+ 2` adds the first line and the extra trailing line.

In an early draft, what I did was to skip all the lines until an empty line is found; unfortunately it does not work with `099_formatting.zig`, causing a test failure.

To summarize again, here are the possible solutions:

  1. Remove trailing whitespace from `Exercise.output`.
      This assumes that there will be only **one** LF at the end of the exercise output.
      This will require to change `099_formatting.zig` to only use one line to separate the rows.

  2. Remove trailing whitespace from `Exercise.output`, but keep one LF in `099_formatting.zig`
      This will make the code inconsistent.

  3. The source of the problem is that the exercise output is unreliable, so keep it reliable.
      Update `ZiglingStep` so that instead of printing the raw output:
      ```zig
      print("{s}PASSED:\n{s}{s}\n", .{ green_text, output, reset_text });
      ```
      print the trimmed output, and add a *mark* using 0x03 (ETX) so that the test can check when the output ends.
     ```zig
      const ETX = 0x03;
      print("{}{s}PASSED:\n{s}{s}\n", .{ETX, green_text, output, reset_text });
      ```
      A possible issue is that the terminal may handle ETX incorrectly (but it should not).
      Additionally, the user may be confused by the fact that the the number of lines does not match what the code do.

**Comment by chrboesch** [comment]
> This will require to change 099_formatting.zig to only use one line to separate the rows.

That isn't an option, because then the exercise looks ugly and is (a bit) loosing the meaning. 
Proposal: You make a new branch with all that changes and I can play around and test it. So it is a bit difficult to understand what works how after these changes. And I need to look for a way for exercise 99.

**Comment by perillo** [comment]
I tried with option 2 and 3, and both works.
Option 3 seems the most reliable (from the test point of view).

This is the code in `build.zig`:
```zig
// Use NUL as a mark for the end of the output.
print("{s}PASSED:\n{s}{s}\n\x00\n", .{ green_text, trimOutput, reset_text });
```

this is the code in `test/tests.zig`:
```zig
        // Skip the exercise output.
        // Use NUL as a mark for the end of the output.
        while (true) {
            const line = try readLine(reader, &buf) orelse @panic("EOF");
            if (line.len == 1 and line[0] == 0x00) return;
        }
```

and this is part of the output of `zig build -Dhealed`:
```
$ zig build -Dhealed
Compiling 098_bit_manipulation2.zig...
Checking 098_bit_manipulation2.zig...
PASSED:
Is this a pangram? true!

Compiling 099_formatting.zig...
Checking 099_formatting.zig...
PASSED:

 X |  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15 
---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
 1 |  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15 

 2 |  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30 

 3 |  3   6   9  12  15  18  21  24  27  30  33  36  39  42  45 

 4 |  4   8  12  16  20  24  28  32  36  40  44  48  52  56  60 

 5 |  5  10  15  20  25  30  35  40  45  50  55  60  65  70  75 

 6 |  6  12  18  24  30  36  42  48  54  60  66  72  78  84  90 

 7 |  7  14  21  28  35  42  49  56  63  70  77  84  91  98 105 

 8 |  8  16  24  32  40  48  56  64  72  80  88  96 104 112 120 

 9 |  9  18  27  36  45  54  63  72  81  90  99 108 117 126 135 

10 | 10  20  30  40  50  60  70  80  90 100 110 120 130 140 150 

11 | 11  22  33  44  55  66  77  88  99 110 121 132 143 154 165 

12 | 12  24  36  48  60  72  84  96 108 120 132 144 156 168 180 

13 | 13  26  39  52  65  78  91 104 117 130 143 156 169 182 195 

14 | 14  28  42  56  70  84  98 112 126 140 154 168 182 196 210 

15 | 15  30  45  60  75  90 105 120 135 150 165 180 195 210 225

Compiling 999_the_end.zig...
Checking 999_the_end.zig...
PASSED:

This is the end for now!
We hope you had fun and were able to learn a lot, so visit us again when the next exercises are available.

```

The problem is that I'm not 100% sure that the NUL character will not cause problems on some terminals.
On Windows, using Windows Terminal and Bash, there were no problems.  But what if the user wants to redirect the output to a file?

**Comment by chrboesch** [comment]
> But what if the user wants to redirect the output to a file?

Then the user opens an issue and we search for a solution. I think nobody can cove all possibiliies. And this one seems a bit unlikely.

**Comment by perillo** [comment]
@chrboesch thanks for the suggestions.

Fortunately using NUL is not necessary, since just the fact that `ZiglingStep` adds only **one** LF after the exercise output solved the problem.

What I plan to do is:
  1. Remove extra trailing whitespace in `Exercise.output`
  2. Document that `Exercise.output` shall not have extra trailing whitespace
  3. Verify that `Exercise.output` does not have extra trailing whitespace using the existing `validate_exercise` function

However I have some doubts #145.
In the first draft, it used the raw exercise output and raw `Exercise.output` string, and checked if the length is the same.
In the final draft, the length check is still here.

I think that, after trimming extra whitespace, we can do an **exact** comparison (using `mem.eql`) instead of a match (using `mem.indexOf`).

**Comment by chrboesch** [comment]
Sounds good!

**Comment by perillo** [comment]
Closed by #250.
